File: backend/app/core/password.py
# app/core/password.py
"""
Password hashing and validation using bcrypt.
Implements secure password hashing with salt and cost factor.
"""
from passlib.context import CryptContext
from app.core.config import settings

# Use bcrypt with a cost factor of 12 (moderate - balance between security and speed)
pwd_context = CryptContext(
    schemes=["bcrypt"],
    deprecated="auto",
    bcrypt__rounds=12
)


def hash_password(password: str) -> str:
    """Hash a plain password using bcrypt."""
    if not password or len(password) < 8:
        raise ValueError("Password must be at least 8 characters")
    return pwd_context.hash(password)


def verify_password(plain_password: str, hashed_password: str) -> bool:
    """Verify a plain password against its bcrypt hash."""
    try:
        return pwd_context.verify(plain_password, hashed_password)
    except Exception:
        return False


def validate_password_strength(password: str) -> list[str]:
    """
    Validate password strength.
    Returns list of error messages (empty if valid).
    
    Requirements:
    - At least 8 characters
    - At least 1 uppercase letter
    - At least 1 lowercase letter
    - At least 1 digit
    - At least 1 special character
    """
    errors = []
    
    if len(password) < 8:
        errors.append("Password must be at least 8 characters long")
    if not any(c.isupper() for c in password):
        errors.append("Password must contain at least one uppercase letter")
    if not any(c.islower() for c in password):
        errors.append("Password must contain at least one lowercase letter")
    if not any(c.isdigit() for c in password):
        errors.append("Password must contain at least one digit")
    if not any(c in "!@#$%^&*()_+-=[]{}|;:,.<>?" for c in password):
        errors.append("Password must contain at least one special character")
    
    return errors

File: frontend/src/utils/password.ts
/**
 * Shared password-strength validation (client-side convenience).
 *
 * The **authoritative** server-side check lives in backend/app/core/password.py.
 * This copy exists solely for realtime UX feedback — never trust it for
 * enforcement.
 */

export function validatePasswordStrength(pw: string): string[] {
  const errors: string[] = [];
  if (pw.length < 8) errors.push("At least 8 characters");
  if (!/[A-Z]/.test(pw)) errors.push("One uppercase letter");
  if (!/[a-z]/.test(pw)) errors.push("One lowercase letter");
  if (!/[0-9]/.test(pw)) errors.push("One digit");
  if (!/[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]/.test(pw))
    errors.push("One special character");
  return errors;
}
